File: 0050-powx-n/0050-powx-n.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        if x == 1:
            return 1
        if x == 0:
            return 0
        if n < 0:
            x = 1/x
            n = abs(n)
            
        if n == 1:
            return x
        if n == 0:
            return 1
        
        powe = self.myPow(x, n//2)
        if n%2 == 0:
            return (powe )**2
        else:
            return x * (powe )**2

File: 0050-powx-n/0050-powx-n.js
/**
 * @param {number} x
 * @param {number} n
 * @return {number}
 */
var myPow = function(x, n) {
    const power = Math.abs(n);

        const recur = (x, n) => {
            if (n === 0) {
                return 1;
            }

            if (n % 2 === 0) {
                const half = recur(x, n / 2);
                return half * half;
            } else {
                const half = recur(x, Math.floor((n - 1) / 2));
                return half * half * x;
            }
        };

        let result = recur(x, power);
        if (n < 0) {
            return 1 / result;
        }
        return result;
};
